### PR TITLE
Give more precise error when disambiguation could not possibly work

### PR DESCRIPTION
**Pull request by lpw25**
Constructor disambiguation can lead to confusing error messages, like:

```ocaml
module M = struct
  type t = A | B | C
  let get _ _ = A 
end

let () =
  match M.get () with
  | A | B | C -> ();;

Line 8, characters 4-5:
8 |   | A | B | C -> ();;
        ^
Error: Unbound constructor A
```

Here, the mistake was clearly missing an argument in the call to `M.get`, but the error message is complaining about an unbound constructor.

This PR adds a more specific error message for cases where disambiguation could not possibly work. For example, the above now gives:

```ocaml
Line 8, characters 4-5:
8 |   | A | B | C -> ();;
        ^
Error: This pattern should not be a constructor, the expected type is
       'a -> M.t
```

### CODE REVIEW

File: typing/ctype.mli [hunk 1]
```diff
@@ -249,8 +249,14 @@ val expand_head_opt: Env.t -> type_expr -> type_expr
...
+  | Typedecl of Path.t * Path.t * type_declaration
+  | Has_no_typedecl
+  | May_have_typedecl
+
 val extract_concrete_typedecl:
-        Env.t -> type_expr -> Path.t * Path.t * type_declaration
+        Env.t -> type_expr -> typedecl_extraction_result
         (* Return the original path of the types, and the first concrete
            type declaration found expanding it.
            Raise [Not_found] if none appears or not a type constructor. *)
```
Comment (ahh) [inline typing/ctype.mli]: Is this comment still accurate?  Looks like it raises less (maybe not at all?)

Comment (lpw25) [reply typing/ctype.mli]: It is not. I'll move it to the type definition and reword it.

File: typing/ctype.ml [hunk 2]
```diff
@@ -1695,18 +1695,29 @@ let _ = forward_try_expand_once := try_expand_safe
...
+        | exception Cannot_expand -> May_have_typedecl
+        | ty ->
+            match extract_concrete_typedecl env ty with
+            | Typedecl(_, p', decl) -> Typedecl(p, p', decl)
+            | Has_no_typedecl -> Has_no_typedecl
+            | May_have_typedecl -> May_have_typedecl
+      end
+  | Tarrow _ | Ttuple _ | Tobject _ | Tfield _ | Tnil
+  | Tvariant _ | Tpackage _ -> Has_no_typedecl
+  | _ -> May_have_typedecl
```
Comment (gasche) [inline typing/ctype.ml]: I would prefer a non-fragile pattern here

File: typing/typecore.ml [hunk 3]
```diff
@@ -65,6 +65,24 @@ type wrong_name = {
...
+  | Boolean
+  | List
+  | Unit
+
+let wrong_kind_sort_of_constructor (lid : Longident.t) =
+  match lid with
+  | Lident "true" | Lident "false" | Ldot(_, "true") | Ldot(_, "false") -> Boolean
+  | Lident "[]" | Lident "::" | Ldot(_, "[]") | Ldot(_, "::") -> List
+  | Lident "()" | Ldot(_, "()") -> Unit
+  | _ -> Constructor
```
Comment (gasche) [inline typing/typecore.ml:85]: I am a bit troubled by the purely syntactic logic here. It is arguably not wrong to think that a user-described datatype that was defined to use `::` as constructor can be referred to as a list, but it is also not exactly correct. I think in similar situations we have argued that the compiler should not embed specialized knowledge of this kind, even for error messages. (But maybe it was for arithmetic operators that are "more library" than those names?)

File: typing/typecore.ml [hunk 4]
```diff
@@ -3059,23 +3101,31 @@ and type_expect_
...
+                Wrong_expected_kind(Record, Expression explanation, ty)
+              else
+                Expr_not_a_record_type ty
             in
-            Some (p0, p, principal)
-          with Not_found -> None
+            raise (Error (loc, env, error))
         in
-        let opath = get_path ty_expected in
+        let opath = get_path loc true ty_expected in
```
Comment (gasche) [inline typing/typecore.ml]: for me an un-labelled boolean argument is a code smell. You could use `get_path loc ~expectation:true ty_expected`.

But maybe there is an even clearer way to distinguish between the two situations here? On a first read I'm not sure I understand what `Expr_not_a_record_type` means compared to `Wrong_expected_kind`, and why we don't have a symmetric case for (non-polymorphic) variant constructors.

Comment (gasche) [reply typing/typecore.ml]: Could this suggest that `type_forcing_context` should be extended with a `Record_project` case or something?

**Review by gasche — COMMENTED** [review]
I haven't done a complete review yet but this looks very promising. The code is easier to follow, the naming is nice, getting rid of exceptions is nice and the error messages are better.

@lpw25, any chance that you would be willing to have a look at #10307? It's in the same broad design space (but less user-facing), I even wondered if the two would conflict (but I don't think so), and it's a fairly small change.

**Comment by Octachron** [comment]
Note that this also seems to be a better alternative to #9395 .

Overall, with type-directed disambiguation in mind, it seems to be a good idea to raise an error early due to a kind mismatch.

File: typing/typecore.ml (hunk 3 above)
Comment (Octachron) [reply typing/typecore.ml:85]: Another weak argument against this specialization is that alternative syntaxes are not required to have `list`, `unit` or `boolean` literals.

Comment (lpw25) [reply typing/typecore.ml:85]: I very deliberately use the terms "list literal", "boolean literal" and "unit literal" in these error messages. These are intended to describe the piece of syntax, not the type that the syntax may result in, and I've tried to stick close to the terminology used for similar constructs in the manual.

Without this specialisation, the messages essentially leak the fact that e.g. `()` is a variant constructor into an error message that is likely seen by beginners. Whereas, an advanced user who knows that you can make a specialised list type, will not be confused by the term "list literal" for e.g. `[a; b; c]`.

Comment (gasche) [reply typing/typecore.ml:85]: I find your argument convincing. The point of @Octachron remains that we are embedding knowledge about the surface syntax in the type-checker, which is unfortunate. A correct long-term fix for this issue would probably be to create explicit parsetree node for those constructs, instead of desugaring in the parsing; but this is orthogonal to the present change.

Is there a simple change that would alleviate this issue? For example, possibly this "literal unparsing" could be moved to a helper function in a `parsing/` file. (I see that `parsing/ast_mapper.ml` already has logic to this effect, that could be shared among several consumers. But then it doesn't use the `Ldot` generalization, so code reuse seems difficult.) I don't see a simple suggestion to make, so I would be fine with leaving things as-is, but maybe @lpw25 or @Octachron have better ideas.

**Comment by lpw25** [comment]
Rebased and addressed review comments.

File: typing/ctype.ml (hunk 2 above)
Comment (lpw25) [reply typing/ctype.ml]: Done

File: typing/typecore.ml (hunk 4 above)
Comment (lpw25) [reply typing/typecore.ml]: This code was getting hard to follow so I refactored it. Should be clearer now.

The difference between the two errors is that one is about using a constructor with an inappropriate expected type, and the other is about using a projection on an expression of the wrong type. You only get the second kind of error for records since there are no projections for variants.

**Review by Octachron — APPROVED** [review]
Raising an error early seems to be the best choice in an environment where type-directed disambiguation is prevalent. In particular, the new error message should avoid few wild-goose chase for unbound constructors (or fields).
 
The code itself is easier to follow.

I still think that the literal handling is a bit adhoc, but this is not the only place where this happens, and I agree that this makes the error message much more understandable  for non-expert users.

**Comment by lpw25** [comment]
Changes entry added -- should be good to merge.
